Diff against tc-proxy

djmitche opened the pull request.

djmitche committed:
turn taskcluster-proxy into relengapi-proxy

This removes the largely-unnecessary godeps support.

This doesn't actually proxy yet -- just compiles

djmitche committed:
convert scopes to perms

djmitche committed:
functional proxy, but no additional auth

djmitche committed:
fetch tmp tokens as necessary

djmitche committed:
refactor, add tests

djmitche committed:
Add docs

lightsofapollo-taskcluster commented on proxy.go:
```diff
@@ -0,0 +1,65 @@
+
+	// httputil's ReverseProxy is not specifically "reverse", and it will
+	// do fine here.  The director transforms outgoing requests.
+	director := func(req *http.Request) {
```
very nice

lightsofapollo-taskcluster commented on relengapi.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+func getTmpToken(url string, issuingToken string, expires time.Time, perms []string) string {
+	// TODO: retry this operation
```
pmoore has some retry code

lightsofapollo-taskcluster commented on relengapi.go:
```diff
@@ -0,0 +1,91 @@
+	}
+
+	client := &http.Client{}
+	reqUrl := fmt.Sprintf("%s/tokenauth/tokens", url)
```
This is probably fine but you may get better url joins using net/url

lightsofapollo-taskcluster commented on relengapi.go:
```diff
@@ -0,0 +1,91 @@
+	}
+
+	if resp.StatusCode != 200 {
+		log.Fatalf("Got '%s' while trying to get new tmp token:\n%s",
```
IMO I would return errors rather then logging/fatal'ing them here

lightsofapollo-taskcluster commented on relengapi.go:
```diff
@@ -0,0 +1,91 @@
+	var responseBody interface{}
+	err = json.Unmarshal(body, &responseBody)
+	if err != nil {
+		log.Fatal(err)
```
^

lightsofapollo-taskcluster commented on taskcluster.go:
```diff
@@ -0,0 +1,20 @@
+
+	task, callSummary := q.Task(taskId)
+	if callSummary.Error != nil {
+		log.Fatalf("Could not fetch taskcluster task '%s' : %s",
```
nit: IMO I would return this error and let the main control flow deal with fatal'ing out

djmitche committed:
return errors from getTmpToken

djmitche committed:
return an error from getToken, and explain why that results in a panic

djmitche committed:
return an error from getTaskScopes, too

djmitche committed:
proxy expects a hostname, not a URL

The proxy doesn't do any path munging -- it just switches on https and
replaces the hostname in the URL.

djmitche committed:
Merge djmitche/relengapi-proxy:host-not-url (PR #2)

djmitche committed:
update README with more usage info

petemoore committed:
Bug 1218344 - retry http requests with exponential backoff

petemoore commented:
Should we close this PR?

djmitche committed:
Merge petemoore/relengapi-proxy:bug1218344 (PR #3)
